### ISSUE TITLE
Add support for Tab and Enter in beta Live Preview

### ISSUE THREAD

**Issue by xl0** [tgrosinger/advanced-tables-obsidian#129]
**Describe the bug**
Tab does not seem to do anything when using the Live Preview mode

**To Reproduce**
1. Install the Obsidian v0.13.4 closed beta.
2. Enable the Live Preview editor mode
3. Navigate to a table, change formatting, press tab
4. The whole row is shifted right instead of the table being re-formatted.

**Expected behavior**
Table re-formatted, cursor moved to the next cell.

**Desktop (please complete the following information):**
 - OS: Ubuntu 20.04.3 LTS
 - Obsidian Version:  v0.13.4
 - Advanced Tables Version: 0.13.0

**Additional context**
I understand that Live Preview is super new and in beta, just a heads up.

**Comment by webdog** [comment]
Same. Seems like I'm experiencing the issues reported in #105 as well.

**Comment by xl0** [comment]
> Same. Seems like I'm experiencing the issues reported in #105 as well.

I don't think it's related, I don't see any errors in the console, the side panel works just fine if I assign a different hotkey it also works fine, just the default Tab and Enter stopped working. Yes, I checked the plugin config, both are enabled. If I assign a different hotkey it also works fine, I just can not use or assign Tab or Enter. Makes me think, it's probably not related to the AT plugin and has something to do with the Obsidian itself.

**Comment by owenfiscus** [comment]
I am also experiencing this issue

**Comment by xl0** [comment]
My understanding is, with Live Preview mode enabled, Obsidian uses the CodeMirror 6 library instead of CodeMirror 5. That's why you have to restart it when changing that option.
It's the open-source JS library for the text/code editor. And CM6 has a completely different way of handling events, which keypresses are, so things no longer work.

And for binding Tab and Enter to hotkeys, that's completely separate from CodeMirror, but it was never an option in Obsidian, you can't bing unmodified Tab or Enter. That's why the plugin had to handle the Tab/Enter keypresses via CodeMrror directly in the first place instead of using the normal hotkey mechanism.

**So, workaround:** Bind something else, like Shift-Tab and Shift-enter via Obsidian hotkeys to the "go to next cell/row" actions.

**Comment by 0dp** [comment]
macOS v. 11.4
Obsidian v. 0.12.19

After activating ATO `enter` and `tab` works as expected

**Comment by tgrosinger** [comment]
> macOS v. 11.4 Obsidian v. 0.12.19
> 
> After activating ATO `enter` and `tab` works as expected

What is "ATO"?

**Comment by 0dp** [comment]
Advanced Tables Obsidian

**Comment by Cutuchiqueno** [comment]
same issue on Windows 11 Obsidian 13.6

**Comment by tgrosinger** [comment]
Tab and Enter are not going to work on any platform right now if you are using the beta Life Preview editor.
No need to add additional error reports.

**Comment by richardlm** [comment]
> 

Just to check @xl0 - I have tried mapping other key combinations to "go to next cell" in Obsidian's Hotkeys, but that still doesn't work when trying to create a table. Did you get that to work?

**Comment by xl0** [comment]
You are right, creating a new table does not seem to work as fluently as it used to, even after binding an extra key. Might be another issue also related to CM6 since it works in CM5 just fine, with the new key. I ended up turning off live mode and just using the Codemirror plugin to get more or less the same result for now.

**Comment by richardlm** [comment]
Quick query @tgrosinger - given that Licat has said that Live Preview is going to be released to the public soon I wondered if you know when you will manage to issue the compatibility update? Many thanks.

**Comment by tgrosinger** [comment]
I am still waiting on Licat to give me APIs to hook this into the editor. There's no way to add this functionality to Live Preview right now.

**Comment by richardlm** [comment]
Ah, ok thanks @tgrosinger. I thought I saw Licat saying the APIs were now open but may have misunderstood.
I don't need to create tables everyday, but when I do it makes me realise just how essential your plug-in is!

**Comment by tgrosinger** [comment]
Yeah, I just checked Discord and Licat has sent me instructions for testing it out. It may be a bit before I get around to figuring out all those changes though.

**Comment by tgrosinger** [comment]
Hey everyone, I have added early support for tab and enter in the Live Preview editor. You will need to have Obsidian 0.13.8 in order to upgrade to it. There are definitely some quirks, for example Obsidian does not seem to be rendering tables in a monospaced font. Also, because it renders links, the number of characters displayed vs the number of characters actually present is going to cause alignment issues.

So TL;DR, I think the plugin works in Live Preview, but it's not yet going to be a smooth experience. I'm going to have to think about how it should behave beyond what's there now.

Please open new issues for any bugs you find.
